llama-index-core/llama_index/core/memory/vector_memory.py

```diff
@@ -0,0 +1,234 @@
+    # so we can delete it when we commit a new node
+    cur_user_msg_key: str = Field(default=CUR_USER_MSG_KEY)
+
+    @validator("vector_index")
+    def validate_vector_index(cls, value: Any) -> Any:
+        """Validate vector index."""
+        # NOTE: we can't import VectorStoreIndex directly due to circular imports,
+        # which is why the type is Any
+        from llama_index.core.indices.vector_store import VectorStoreIndex
+
+        if not isinstance(value, VectorStoreIndex):
+            raise ValueError(
+                f"Expected 'vector_index' to be an instance of VectorStoreIndex, got {type(value)}"
+            )
+        return value
+
+    @classmethod
+    def class_name(cls) -> str:
+        """Get class name."""
+        return "VectorMemory"
+
+    @classmethod
+    def from_defaults(
+        cls,
+        vector_store: Optional[VectorStore] = None,
+        embed_model: Optional[EmbedType] = None,
+        index_kwargs: Optional[Dict] = None,
+        retriever_kwargs: Optional[Dict] = None,
+    ) -> "VectorMemory":
+        """Create vector memory.
+
+        Args:
+            vector_store (Optional[VectorStore]): vector store
+            embed_model (Optional[EmbedType]): embedding model
+            index_kwargs (Optional[Dict]): kwargs for initializing the index
+            retriever_kwargs (Optional[Dict]): kwargs for initializing the retriever
+
+        """
+        from llama_index.core.indices.vector_store import VectorStoreIndex
+
+        index_kwargs = index_kwargs or {}
+        retriever_kwargs = retriever_kwargs or {}
+
+        if vector_store is None:
+            # initialize a blank in-memory vector store
+            # NOTE: can't easily do that from `from_vector_store` at the moment.
+            index = VectorStoreIndex.from_documents(
+                [], embed_model=embed_model, **index_kwargs
+            )
+        else:
+            index = VectorStoreIndex.from_vector_store(
+                vector_store, embed_model=embed_model, **index_kwargs
+            )
+        return cls(vector_index=index, retriever_kwargs=retriever_kwargs)
+
+    def get(
+        self, input: Optional[str] = None, initial_token_count: int = 0, **kwargs: Any
+    ) -> List[ChatMessage]:
+        """Get chat history."""
+        if input is None:
+            raise ValueError("Input must be provided to get chat history.")
+
+        # retrieve from index
+        retriever = self.vector_index.as_retriever(**self.retriever_kwargs)
+        nodes = retriever.retrieve(input or "")
+
+        # retrieve underlying messages
+        messages = [
+            ChatMessage.parse_obj(sub_dict)
+            for node in nodes
+            for sub_dict in node.metadata["sub_dicts"]
+        ]
+
+        # add system message
+        if self.system_message:
+            messages = [
+                ChatMessage.from_str(self.system_message, role=MessageRole.SYSTEM)
```

Comment:
Hmm, I'm not sure this makes sense to do? This will end up duplicating system prompts in certain chat histories, and a lot of APIs don't like that (openai seems to be the exception).

It makes more sense to include the "system message" as a preamble for when return_single_message is true?

I wonder what the motivation for this was @jerryjliu ? (or maybe it was @nerdai that wrote this, not sure)

Reply:
wasn't me, but going to review a bit more closely and let you know what I think as well

Reply:
Here it is. The "system" message serves as a sort of context setting for the conversation flow. Otherwise, the LLM might get confused with the flow of the conversation.

![image](https://github.com/run-llama/llama_index/assets/92402603/ac481d19-4268-4a65-a14a-84890261c585)

Reply:
This will certainly duplicate system messages every time we retrieve from longer-term history, which as you say may pose some issues with some APIs. I sort of think the default for `return_single_message` should be True and we should use a system message in this case.

Maybe this type of prompt will lead to better results... not sure and probably should be tested. At least from a retrieve conversation perspective, a human will be able to analyze it more easily since its clear what is actual conversation versus what is just a recalled "playback" of a past dialogue.

Reply:
I think
a) we should not be inserting a system prompt by default
b) return single message should probably be true by default. But the current formatting may be confusing (single content string, no assistant/user/tool prefixes, no check for empty messages)
b) If we have pre-amble (like the existing system message), we can add it as part of of the single message blob too

Reply:
oops sorry just seeing this. i actually didn't know that having multiple sysprompts might break LLM APIs (have mostly tested against openai)